src/transformers/generation/flax_utils.py

```diff
@@ -997,8 +1028,13 @@ def beam_search_body_fn(state, input_ids_length=1):
         sequences = jnp.where(none_finished[:, None, None], state.sequences, state.running_sequences)
         scores = jnp.where(none_finished[:, None], state.scores, state.running_scores)
 
-        # take best beam for each batch
-        sequences = sequences[:, 0]
-        scores = scores[:, 0]
+        if output_scores:
+            # take all beam of each batch
+            sequences = sequences[:, :]
+            scores = scores[:, :]
```

Comment:
My worry is now that we have three different kinds of functionality for each of our generation methods:
1. Greedy: `scores` is a tensor of shape (bsz, seq len, vocab size)
2. Sampling: we don't have the option of outputting the scores (no option for `output_scores`)
3. Beam: `scores` is a tensor of shape (bsz, seq len, beams) if `output_scores=True`, otherwise (bsz, seq len). This is more like `sequence_scores` in PyTorch: https://huggingface.co/docs/transformers/internal/generation_utils#transformers.generation.BeamSearchDecoderOnlyOutput.sequences_scores. `sequences` also changes shape depending on the value of `output_scores`. 

=> we have quite a lot of inconsistency between these sampling methods. Ideally, the signature of these methods should all be the same, such that they all accept `output_scores` and are consistent in their outputs regardless of the value of `output_scores` (e.g. beam always returns either top beam or all beams).

Reply:
I think it is a huge goal to make flax models output completely consistent with Pytorch model's output, while Flax codes (i.e. `src/transformers/generation/flax_utils.py`) has a different structure and simpler implementation rather than PyTorch codes (i.e. `src/transformers/generation/utils.py`)
I updated the code to make flax more consistent with PyTorch, by changing the name of `scores` to `sequences_scores` for BeamSearchOutput and return it when `output_scores=True`
Besides, in pytorch, both `sequences` and `sequences_scores` are returning the top `num_return_sequences` (not just the first one). So the Flax code is modified to return the top `num_return_sequences` sequences instead of the first or all `num_beams`
For adding scores to BeamSearchOutput, I suggest to throw it as a TODO for next updates.
About Sampling, I also suggest it to be a TODO for future.

Please refer to https://github.com/huggingface/transformers/commit/f0e35070e8b40b4a1cc8e23cf00c6e008f94989c

Reply:
Awesome! Thanks for the details - agree with the decisions you've made here